feature(fixtures): Load fixtures into the database

mrapacz opened the pull request:
**Summary**
Add fixtures module responsible for loading previously prepared questions. 

**Related issues**
Closes: #14 

**Test plan**
Run `mix ecto.migrate` to update the database schema.
`pip install -r requirements.txt` to install modules required for the fixtures module to work.
`python main.py -h` to learn how to run the module.

mrapacz committed:
enhancement(aion): Generate Subject, Question and Answer models

Generate models using phoenix.gen.json

mrapacz committed:
enhancement(aion): Generate Subject, Question and Answer models

Generate models using phoenix.gen.json

mrapacz committed:
fix(web/router.ex): Pipe resources through :api

Pipe resources through :api to omit pointing to html version of the
controller.

mrapacz committed:
fix(web/router.ex): Pipe resources through :api

Pipe resources through :api to omit pointing to html version of the
controller.

mrapacz committed:
Add .gitignore file

mrapacz committed:
Add .gitignore file

mrapacz committed:
feature(fixtures): Introduce question parser

Introduce a Python module enabling one to parse files containing
questions to Python dict objects.

mrapacz committed:
feature(fixtures): Introduce question parser

Introduce a Python module enabling one to parse files containing
questions to Python dict objects.

mrapacz committed:
feature(fixtures): Fix Polish notation

Replace incorrect occurrences of Polish characters with the proper ones
e.g. '%s' -> 'ś'

mrapacz committed:
feature(fixtures): Fix Polish notation

Replace incorrect occurrences of Polish characters with the proper ones
e.g. '%s' -> 'ś'

mrapacz committed:
fix(fixtures): Fix naming

Switch from questionparser's parser to get_questions and from loader.py
to main.py

mrapacz committed:
fix(fixtures): Fix naming

Switch from questionparser's parser to get_questions and from loader.py
to main.py

mrapacz committed:
Add readme and requirements file

mrapacz committed:
Add readme and requirements file

mrapacz committed:
feat(fixtures): Enable loading subjects into db

Enable loading subject into the database

mrapacz committed:
feat(fixtures): Enable loading subjects into db

Enable loading subject into the database

mrapacz committed:
fix(fixtures): Fix regex to match images

Add missing group for matching optional images linked to questions

mrapacz committed:
fix(fixtures): Fix regex to match images

Add missing group for matching optional images linked to questions

mrapacz committed:
fix(fixtures): Allow multilines in the questions

Fix the regex pattern allowing one to write longer (multiline)
questions

mrapacz committed:
fix(fixtures): Allow multilines in the questions

Fix the regex pattern allowing one to write longer (multiline)
questions

mrapacz committed:
fix(fixtures): Include trailing whitespace in regex

Sometimes the questions don't finish with a newline character but with
an additional space

mrapacz committed:
fix(fixtures): Include trailing whitespace in regex

Sometimes the questions don't finish with a newline character but with
an additional space

mrapacz committed:
feature(fixtures): Insert questions into db

Add queries responsible for inserting questions into the database

mrapacz committed:
feature(fixtures): Insert questions into db

Add queries responsible for inserting questions into the database

mrapacz committed:
fix(elixir): Change question content to text field

Change the type of question content to text as some of the questions
exceeded varchar(255)

mrapacz committed:
fix(elixir): Change question content to text field

Change the type of question content to text as some of the questions
exceeded varchar(255)

mrapacz committed:
fix(fixtures): Move queries to a separate module

mrapacz committed:
fix(fixtures): Move queries to a separate module

mrapacz committed:
feature(fixtures): Insert answers into the db

Add functions responsible for loading all of the answers into the
database

mrapacz committed:
feature(fixtures): Insert answers into the db

Add functions responsible for loading all of the answers into the
database

mrapacz committed:
temp

mrapacz committed:
merge

pmrukot commented on fixtures/questionparser.py:
```diff
@@ -0,0 +1,43 @@
+            question = {group_name: match.group(group_name) for group_name in group_names if match.group(group_name)}
+            question = fix_coding(question)
+            questions.append(question)
+            assert 'answers' in question and 'subject' in question and 'question' in question
```
assert seems to fail in some cases which breaks migration completely, possible workaround?

mrapacz replied:
If the assert doesn't work that means that the questions do not hold to the requested format e.g. there were no answers specified. There shouldn't be any issues with the latest version of questions database, I tested it and it runs smoothly

pmrukot commented on fixtures/queries/answers.py:
```diff
@@ -0,0 +1,37 @@
+
+
+def insert_answers(question, conn):
+    answers = question['answers'].split(';')
```
Got "KeyError" while migrating which broke the process. Could we use handle it somehow, for example when question has no answers we abort the insertion?

mrapacz replied:
Actually, the assert statement in the questionparser takes care of a situation like that. It shouldn't appear with the latest question fixtures we've got anyway.

pmrukot commented on fixtures/queries/questions.py:
```diff
@@ -0,0 +1,36 @@
+        else:
+            query = """SELECT id FROM questions WHERE content = '{}' AND image_name = {} AND subject_id = '{}'""".format(
+                content, image, subject_id)
+        cur.execute(query)
```
Got:
```
psycopg2.DataError: value too long for type character varying(255)
```
We may want to handle errors like that or setup our database tables properly.

mrapacz replied:
That's what the latest migration was about, I decided to change the question content field to text field instead of varchar(255). It works great with the latest version of the fixtures.

pmrukot reviewed (changes requested):
I've managed to migrate the questions successfully but  there are some issues that needs to resolved (pointed out in a comments) to make it work properly. 

Also @jtkpiotr could look into docker related stuff and try to create a proper database config so it actually works.

It's also worth noticing that Ecto creates `aion_dev` and `aion_test` databases for Phoenix. We need to work on sane configuration and using environment variables. I guess the `config.py` file should have the config from `aion/config/dev.exs` file.

pmrukot commented:
I also suggest moving `codingfix.py`, `config.py`, `loader.py` and `questionparser.py` to `/fixtures/src` directory and leave only `main.py` in `/fixtures`

mrapacz committed:
fix(aion): Create scripts/script.sh

Add a script responsible for running ecto.migrate, loading the fixtures
into the database and running the phoenix server

mrapacz committed:
fix(fixtures): Move python files to /fixtures/src/

Move python source files to /fixtures/src and fix imports

mrapacz commented:
@pmrukot The things related to configs are not to be changed as they point at the database that will be created by docker and they are configured well, moving everything to docker was what I needed to do. I moved the python files to /fixtures/src as requested. I also fixed the docker configuration and right now it should set up everything without any issues. Make sure you use the latest version of fixtures for testing as the previous ones were often badly formatted and caused problems related to what you wrote about in your review.

jtkpiotr commented on Dockerfile:
```diff
@@ -13,4 +15,10 @@ RUN apt-get update \
+
+# Install python pip and dependencies
+RUN apt-get -y install python3-pip \
+    && pwd \
```
why pwd and ls here :)?

jtkpiotr commented on scripts/script.sh:
```diff
@@ -0,0 +1,4 @@
+sleep 4
+mix ecto.migrate
+python3 /fixtures/main.py -path /fixtures/jpks/
```
I guess we want to populate database only once. We don't want to run this script whenever app is starting.

pmrukot commented:
Move `fixtures/.gitignore` content to root `.gitgnore`

pmrukot commented:
Closing via #18
